File: calibration/calibration_SNOWPACK/dsnow_parameter_optimization_DE.R
##############################################################################
# DELTASNOW PARAMETER OPTIMIZATION (Differential Evolution)
##############################################################################
#
# Objective:
#   final_score = w1 * NRMSE_SWE + w2 * NRMSE_rho + w3 * NBIAS_SWE
#
# where:
#   NRMSE_SWE  = RMSE_SWE  / mean_obs_SWE
#   NRMSE_rho  = RMSE_rho  / mean_obs_rho
#   NBIAS_SWE  = |bias_SWE| / mean_obs_SWE
#
# Weights (w1, w2, w3) can be set below.
#
##############################################################################

library(DEoptim)
library(zoo)
library(foreach)
library(doParallel)
library(lubridate)
library(nixmass)
library(tidyverse)

# ----------------------------------------------------------------------------
# CONFIGURATION (USER-ADJUSTABLE)
# ----------------------------------------------------------------------------

# Weights for the combined objective (should sum to 1)
WEIGHT_SWE_NRMSE  <- 0.2    # weight for normalized SWE RMSE
WEIGHT_RHO_NRMSE  <- 0.8    # weight for normalized density RMSE
WEIGHT_SWE_NBIAS  <- 0.0    # weight for normalized SWE absolute bias

# Override weights from command-line: Rscript script.R <w_swe> <w_rho> <w_bias>
.args <- commandArgs(trailingOnly = TRUE)
if (length(.args) == 3) {
  WEIGHT_SWE_NRMSE <- as.numeric(.args[1])
  WEIGHT_RHO_NRMSE <- as.numeric(.args[2])
  WEIGHT_SWE_NBIAS <- as.numeric(.args[3])
}
cat(sprintf("Weights: SWE_NRMSE=%.3f  RHO_NRMSE=%.3f  SWE_NBIAS=%.3f\n",
            WEIGHT_SWE_NRMSE, WEIGHT_RHO_NRMSE, WEIGHT_SWE_NBIAS))

# Data settings
season_start   <- "-08-01"       # start of hydrological year (month-day)
season_end     <- "-07-31"       # end of hydrological year
start_of_block <- 8              # month number for block assignment

# Model settings
EPS <- 1e-6                      # threshold for snow depth in density calculation

# DE settings
DE_ITERMAX   <- 100              # maximum number of DE iterations
DE_NP        <- 70               # population size (rule of thumb: 10 * n_params)
DE_F         <- 0.8              # differential weight
DE_CR        <- 0.9              # crossover probability
DE_STRATEGY  <- 2                # 1 = DE/rand/1/bin, 2 = DE/local-to-best/1/bin

# ----------------------------------------------------------------------------
# LOAD OBSERVATIONAL DATA
# ----------------------------------------------------------------------------
# Adjust the path to your .rda file containing a list of zoo objects named 'd_obs'
d_obs <- get(load(
  "/Users/jakobwerkgarner/code/mt_dsnow/calibration/calibration_SNOWPACK/data/d_obs_SNOWPACK.rda"
))

# ----------------------------------------------------------------------------
# HELPER FUNCTIONS
# ----------------------------------------------------------------------------

#' Assign a hydrological season (year) to a date
set_season <- function(date, start_month = 8) {
  date <- as.Date(date)
  yr <- year(date)
  mo <- month(date)
  ifelse(mo < start_month, yr - 1, yr)
}

#' Root Mean Square Error
rmse <- function(obs, mod) {
  ok <- is.finite(obs) & is.finite(mod)
  if (!any(ok)) return(NA_real_)
  sqrt(mean((mod[ok] - obs[ok])^2))
}

#' Normalized RMSE
nrmse <- function(obs, mod) {
  ok <- is.finite(obs) & is.finite(mod)
  if (!any(ok)) return(NA_real_)
  rmse_val <- sqrt(mean((mod[ok] - obs[ok])^2))
  mean_obs <- mean(obs[ok], na.rm = TRUE)
  rmse_val / mean_obs
}

#' Normalized Absolute Bias
nbias <- function(obs, mod) {
  ok <- is.finite(obs) & is.finite(mod)
  if (!any(ok)) return(NA_real_)
  bias_val <- mean(mod[ok] - obs[ok], na.rm = TRUE)
  mean_obs <- mean(obs[ok], na.rm = TRUE)
  abs(bias_val) / mean_obs
}

#' Combined score: weighted sum of NRMSE_SWE, NRMSE_rho, and NBIAS_SWE
combined_score <- function(df, eps = EPS) {
  swe_obs <- df$swe_obs
  swe_mod <- df$swe_mod
  hs <- df$hs
  
  # SWE metrics
  nrmse_swe <- nrmse(swe_obs, swe_mod)
  nbias_swe <- nbias(swe_obs, swe_mod)
  
  # Bulk density metrics (only where snow depth > eps)
  rho_obs <- ifelse(is.finite(hs) & hs > eps, swe_obs / hs, NA_real_)
  rho_mod <- ifelse(is.finite(hs) & hs > eps, swe_mod / hs, NA_real_)
  nrmse_rho <- nrmse(rho_obs, rho_mod)
  
  # Weighted combination
  score <- WEIGHT_SWE_NRMSE * nrmse_swe +
           WEIGHT_RHO_NRMSE * nrmse_rho +
           WEIGHT_SWE_NBIAS * nbias_swe
  
  # Attach detailed metrics for verbose printing
  attr(score, "metrics") <- list(
    rmse_swe  = rmse(swe_obs, swe_mod),
    rmse_rho  = rmse(rho_obs, rho_mod),
    nrmse_swe = nrmse_swe,
    nrmse_rho = nrmse_rho,
    bias_swe  = mean(swe_mod - swe_obs, na.rm = TRUE),
    nbias_swe = nbias_swe,
    mean_swe  = mean(swe_obs[is.finite(swe_obs) & is.finite(swe_mod)], na.rm = TRUE),
    mean_rho  = mean(rho_obs[is.finite(rho_obs) & is.finite(rho_mod)], na.rm = TRUE)
  )
  
  return(score)
}

# ----------------------------------------------------------------------------
# SPLIT DATA INTO FIT AND VALIDATION (even/odd years per station)
# ----------------------------------------------------------------------------
d_obs_fit <- list()
d_obs_val <- list()

for (station in names(d_obs)) {
  cat("Processing station:", station, "\n")
  
  d <- d_obs[[station]]
  years <- unique(year(index(d)))
  
  fit_list <- list()
  val_list <- list()
  counter <- 1
  
  for (y in years[1:(length(years) - 1)]) {
    winter <- subset(
      d,
      index(d) >= as.Date(paste0(y, season_start)) &
        index(d) < as.Date(paste0(y + 1, season_end)) + 1
    )
    
    if (nrow(winter) < 200) {
      cat("  skipping year", y, "- only", nrow(winter), "values\n")
      next
    }
    
    if (
      nrow(winter) < 365 &&
      (as.numeric(winter$Hobs[1]) > 0.05 ||
       as.numeric(winter$Hobs[nrow(winter)]) > 0.05)
    ) {
      cat("  skipping year", y, "- incomplete winter with snow at edge\n")
      next
    }
    
    if (counter %% 2 == 0) {
      fit_list[[length(fit_list) + 1]] <- winter
    } else {
      val_list[[length(val_list) + 1]] <- winter
    }
    counter <- counter + 1
  }
  
  d_obs_fit[[station]] <- if (length(fit_list) > 0) do.call(rbind, fit_list) else NULL
  d_obs_val[[station]] <- if (length(val_list) > 0) do.call(rbind, val_list) else NULL
}

# ----------------------------------------------------------------------------
# PREPARE FIT DATA (convert zoo to tibble)
# ----------------------------------------------------------------------------
fit_data_list <- list()

for (station in names(d_obs_fit)) {
  x <- d_obs_fit[[station]]
  if (is.null(x) || length(x) == 0) next
  
  dat <- tryCatch(as_tibble(coredata(x)), error = function(e) NULL)
  if (is.null(dat) || ncol(dat) == 0) next
  
  if (ncol(dat) == 2 && all(is.na(colnames(dat)))) {
    colnames(dat) <- c("Hobs", "SWEobs")
  }
  
  if (!all(c("Hobs", "SWEobs") %in% colnames(dat))) {
    cat("Skipping station", station, "- missing Hobs or SWEobs\n")
    next
  }
  
  fit_data_list[[station]] <- tibble(
    date    = as.Date(index(x)),
    name    = station,
    hs      = dat$Hobs,
    swe_obs = dat$SWEobs,
    block   = set_season(index(x), start_of_block)
  )
}

d_obs_fit_tibble <- bind_rows(fit_data_list)

# ----------------------------------------------------------------------------
# OBJECTIVE FUNCTION (called by DEoptim)
# ----------------------------------------------------------------------------
# Note: DEoptim passes parameters as a vector directly (no scaling needed internally
# if we define bounds in real space). We keep verbose output consistent.

minimize_score_de <- function(par, data, verbose = FALSE) {
  
  # Ensure a scalar numeric is returned even on error
  result <- tryCatch({
    
    if (verbose) {
      param_names <- c("rho.max", "rho.null", "c.ov", "k.ov", "k", "tau", "eta.null")
      values <- paste(
        paste0(param_names, "=", round(par, 6)),
        collapse = ", "
      )
      cat("params =", values, "\n")
    }
    
    # Run model for all stations and blocks in parallel
    station_results <- foreach(
      station = unique(data$name),
      .packages = c("dplyr", "tidyr", "nixmass"),
      .combine = bind_rows
    ) %dopar% {
      
      data_station <- data %>% filter(name == station)
      blocks <- unique(data_station$block)
      
      block_results <- lapply(blocks, function(b) {
        df_block <- data_station %>% filter(block == b)
        
        full_dates <- tibble(
          date = seq(min(df_block$date), max(df_block$date), by = "1 day")
        )
        
        joined <- df_block %>%
          select(date, hs, swe_obs) %>%
          right_join(full_dates, by = "date") %>%
          arrange(date)
        
        swe_mod <- tryCatch(
          {
            joined %>%
              select(date, hs) %>%
              mutate(date = as.character(date)) %>%
              nixmass::swe.delta.snow(
                model_opts = list(
                  rho.max  = par[1],
                  rho.null = par[2],
                  c.ov     = par[3],
                  k.ov     = par[4],
                  k        = par[5],
                  tau      = par[6],
                  eta.null = par[7]
                ),
                dyn_rho_max = FALSE
              )
          },
          error = function(e) rep(NA_real_, nrow(joined))
        )
        
        joined %>%
          mutate(swe_mod = swe_mod) %>%
          drop_na()
      })
      
      bind_rows(block_results)
    }
    
    dff <- station_results
    
    if (nrow(dff) == 0) {
      cat("No valid model output. Returning large penalty.\n")
      return(1e12)
    }
    
    score_with_attr <- combined_score(dff)
    metrics <- attr(score_with_attr, "metrics")
    score <- as.numeric(score_with_attr)   # strip attributes
    
    if (!is.finite(score)) {
      cat("Score is not finite. Returning large penalty.\n")
      return(1e12)
    }
    
    # Verbose output
    if (verbose && !is.null(metrics)) {
      cat(
        "| bias_swe =", round(metrics$bias_swe, 4),
        "| RMSE_SWE =", round(metrics$rmse_swe, 4),
        "| NRMSE_SWE =", round(metrics$nrmse_swe, 4),
        "| NBIAS_SWE =", round(metrics$nbias_swe, 4),
        "| RMSE_rho =", round(metrics$rmse_rho, 4),
        "| NRMSE_rho =", round(metrics$nrmse_rho, 4),
        "| final_score =", round(score, 4),
        "\n"
      )
    } else if (verbose) {
      cat("Score =", round(score, 4), "\n")
    }
    
    return(score)
    
  }, error = function(e) {
    cat("Error in objective function:", e$message, "\n")
    return(1e12)
  })
  
  # Final safeguard
  if (length(result) != 1 || !is.numeric(result)) {
    return(1e12)
  }
  return(result)
}

# ----------------------------------------------------------------------------
# PARAMETER BOUNDS (in real/unscaled space)
# ----------------------------------------------------------------------------
# Define reasonable lower and upper bounds for each parameter

par_lower <- c(
  rho.max  = 300,
  rho.null = 60,
  c.ov     = 1e-6,
  k.ov     = 0.01,
  k        = 0.001,
  tau      = 0.001,
  eta.null = 1e5
)

par_upper <- c(
  rho.max  = 600,
  rho.null = 150,
  c.ov     = 0.01,
  k.ov     = 1.0,
  k        = 0.1,
  tau      = 0.1,
  eta.null = 1e8
)

cat("\nParameter bounds:\n")
cat("Lower:", par_lower, "\n")
cat("Upper:", par_upper, "\n")

# ----------------------------------------------------------------------------
# PARALLEL SETUP
# ----------------------------------------------------------------------------
nc <- parallel::detectCores(logical = TRUE) - 1
nc <- max(1, nc)
cl <- parallel::makeCluster(nc)
doParallel::registerDoParallel(cl)

# Export required objects to cluster for nested parallelism
parallel::clusterExport(cl, c("WEIGHT_SWE_NRMSE", "WEIGHT_RHO_NRMSE", 
                               "WEIGHT_SWE_NBIAS", "EPS"))

# ----------------------------------------------------------------------------
# TEST OBJECTIVE FUNCTION AT INITIAL GUESS
# ----------------------------------------------------------------------------
par_init <- c(
  rho.max  = 401.2588,
  rho.null = 81.19417,
  c.ov     = 0.0005104722,
  k.ov     = 0.37856737,
  k        = 0.02993175,
  tau      = 0.02362476,
  eta.null = 8523356
)

cat("\nTesting objective function at initial guess...\n")
test_score <- minimize_score_de(
  par     = par_init,
  data    = d_obs_fit_tibble,
  verbose = TRUE
)
cat("Initial score =", test_score, "\n")

# ----------------------------------------------------------------------------
# OPTIMIZATION (Differential Evolution)
# ----------------------------------------------------------------------------
cat("\nStarting Differential Evolution optimization...\n")
cat("  itermax =", DE_ITERMAX, "\n")
cat("  NP =", DE_NP, "\n")
cat("  F =", DE_F, "\n")
cat("  CR =", DE_CR, "\n")
cat("  strategy =", DE_STRATEGY, "\n\n")

opt <- DEoptim(
  fn      = minimize_score_de,
  lower   = par_lower,
  upper   = par_upper,
  data    = d_obs_fit_tibble,
  verbose = TRUE,
  control = DEoptim.control(
    itermax   = DE_ITERMAX,
    NP        = DE_NP,
    F         = DE_F,
    CR        = DE_CR,
    strategy  = DE_STRATEGY,
    trace     = 10,           # print progress every 10 iterations
    parallelType = 0          # we handle parallelism inside the objective
  )
)

# Extract best parameters
best_par <- opt$optim$bestmem
names(best_par) <- names(par_lower)
best_value <- opt$optim$bestval

cat("\n======================================\n")
cat("Optimization complete!\n")
cat("Best score:", best_value, "\n")
cat("Best parameters:\n")
print(best_par)

# Print final parameters as Python pydeltasnow call
fmt <- function(x) format(x, scientific = FALSE, trim = TRUE, digits = 10)

cat(
  "\nswe_results = pydeltasnow.swe_deltasnow(\n",
  "    idata,\n",
  "    rho_max   = ", fmt(best_par["rho.max"]), ",\n",
  "    rho_null  = ", fmt(best_par["rho.null"]), ",\n",
  "    c_ov      = ", fmt(best_par["c.ov"]), ",\n",
  "    k_ov      = ", fmt(best_par["k.ov"]), ",\n",
  "    k         = ", fmt(best_par["k"]), ",\n",
  "    tau       = ", fmt(best_par["tau"]), ",\n",
  "    eta_null  = ", fmt(best_par["eta.null"]), ",\n",
  "    hs_input_unit=\"m\",\n",
  "    swe_output_unit=\"mm\",\n",
  "    output_series_name=\"SWE_mod\"\n",
  ")\n",
  sep = ""
)

# ----------------------------------------------------------------------------
# SAVE RESULTS
# ----------------------------------------------------------------------------

make_weight_tag <- function(x) {
  out <- format(x, scientific = FALSE, trim = TRUE, digits = 6)
  out <- gsub("\\.", "p", out)
  out <- gsub("-", "m", out)
  out
}

weight_vals <- c(
  SWE_NRMSE = WEIGHT_SWE_NRMSE,
  RHO_NRMSE = WEIGHT_RHO_NRMSE,
  SWE_NBIAS = WEIGHT_SWE_NBIAS
)

weight_tag <- paste0(
  names(weight_vals), "_", make_weight_tag(weight_vals),
  collapse = "__"
)

save_dir <- "/Users/jakobwerkgarner/code/mt_dsnow/calibration/calibration_SNOWPACK/data/R_opt_logs_DE"
dir.create(save_dir, recursive = TRUE, showWarnings = FALSE)

save_file <- file.path(
  save_dir,
  paste0("opt_results_DE__", weight_tag, ".rds")
)

saveRDS(opt, file = save_file)

cat("\nOptimization finished. Results saved to:\n", save_file, "\n", sep = "")

# ----------------------------------------------------------------------------
# STOP CLUSTER
# ----------------------------------------------------------------------------
parallel::stopCluster(cl)

File: calibration/archieve/dsnow_parameter_optimization_DE.py
##############################################################################
# DELTASNOW PARAMETER OPTIMIZATION FROM SMET FILES - DIFFERENTIAL EVOLUTION
# with FutureWarning suppression inside parallel workers
##############################################################################

import sys
import contextlib
import io
import warnings

# Suppress FutureWarnings globally (for the main process)
warnings.filterwarnings("ignore", category=FutureWarning)

import numpy as np
import pandas as pd
from pathlib import Path
from scipy.optimize import differential_evolution
from joblib import Parallel, delayed

sys.path.insert(0, "/Users/jakobwerkgarner/code/mt_dsnow/snow_to_swe_master")
from main import SnowToSwe

##############################################################################
# PATHS AND SETTINGS
##############################################################################

INPUT_DIR = Path("/Users/jakobwerkgarner/code/mt_dsnow/par_sens/SNOWPACK_data")

SEASON_START_MONTH = 8
MIN_WINTER_LENGTH = 200
EPS = 1e-6
N_JOBS = -1

##############################################################################
# OBJECTIVE WEIGHTS  (should sum to 1)
##############################################################################

WEIGHT_SWE_NRMSE = 0.5   # normalized RMSE of SWE
WEIGHT_RHO_NRMSE = 0.5   # normalized RMSE of bulk density
WEIGHT_SWE_NBIAS = 0.0   # normalized absolute bias of SWE

##############################################################################
# VARIABLES
##############################################################################

HS_COL = "HS_meas"
SWE_COL = "SWE"

##############################################################################
# SMET FUNCTIONS
##############################################################################

def read_smet(file_path):
    header = []
    fields = None
    data_rows = []
    data_start = False

    with open(file_path, "r") as f:
        for line in f:
            line = line.rstrip()

            if line.startswith("fields"):
                fields = line.split("=", 1)[1].strip().split()

            elif line.strip() == "[DATA]":
                data_start = True
                continue

            elif not data_start:
                header.append(line)

            else:
                if line:
                    data_rows.append(line.split())

    df = pd.DataFrame(data_rows, columns=fields)
    df["timestamp"] = pd.to_datetime(df["timestamp"])

    return header, df


def extract_station_id(header):
    for line in header:
        if line.startswith("station_id"):
            return line.split("=", 1)[1].strip()
    raise ValueError("station_id not found in SMET header")


##############################################################################
# GENERAL HELPER FUNCTIONS
##############################################################################

def rmse(obs, mod):
    obs = np.asarray(obs, dtype=float)
    mod = np.asarray(mod, dtype=float)

    ok = np.isfinite(obs) & np.isfinite(mod)
    if not np.any(ok):
        return np.nan

    return np.sqrt(np.mean((obs[ok] - mod[ok]) ** 2))


def combined_score(df, eps=1e-6):
    swe_obs = df["swe_obs"].to_numpy()
    swe_mod = df["swe_mod"].to_numpy()
    hs      = df["hs"].to_numpy()

    # Bulk density (only where snow is present)
    rho_obs = np.where(hs > eps, swe_obs / hs, np.nan)
    rho_mod = np.where(hs > eps, swe_mod / hs, np.nan)

    rmse_swe = rmse(swe_obs, swe_mod)
    rmse_rho = rmse(rho_obs, rho_mod)

    mean_swe = np.nanmean(swe_obs)
    mean_rho = np.nanmean(rho_obs)

    ok = np.isfinite(swe_obs) & np.isfinite(swe_mod)
    bias_swe = np.mean(swe_mod[ok] - swe_obs[ok]) if np.any(ok) else np.nan

    nrmse_swe = rmse_swe / mean_swe  if (np.isfinite(rmse_swe) and mean_swe > 0) else np.inf
    nrmse_rho = rmse_rho / mean_rho  if (np.isfinite(rmse_rho) and mean_rho > 0) else np.inf
    nbias_swe = abs(bias_swe) / mean_swe if (np.isfinite(bias_swe) and mean_swe > 0) else np.inf

    score = (
        WEIGHT_SWE_NRMSE * nrmse_swe +
        WEIGHT_RHO_NRMSE * nrmse_rho +
        WEIGHT_SWE_NBIAS * nbias_swe
    )
    return score if np.isfinite(score) else np.inf


def is_valid_winter(df):
    if len(df) < MIN_WINTER_LENGTH:
        return False

    if len(df) < 365:
        first_hs = float(df["hs"].iloc[0])
        last_hs = float(df["hs"].iloc[-1])
        if first_hs > 0.05 or last_hs > 0.05:
            return False

    return True


##############################################################################
# LOAD ALL SMET FILES
##############################################################################

def load_all_smet_data(input_dir):
    all_data = []

    for smet_file in input_dir.glob("*.smet"):
        print(f"Reading {smet_file.name}")

        header, df = read_smet(smet_file)
        station_id = extract_station_id(header)

        needed = ["timestamp", HS_COL, SWE_COL]
        missing = [col for col in needed if col not in df.columns]
        if missing:
            raise ValueError(f"{smet_file.name} is missing columns: {missing}")

        df = df[needed].copy()

        df[[HS_COL, SWE_COL]] = (
            df[[HS_COL, SWE_COL]]
            .apply(pd.to_numeric, errors="coerce")
            .replace(-999, np.nan)
        )

        df_daily = (
            df.set_index("timestamp")[[HS_COL, SWE_COL]]
            .resample("D")
            .mean()
            .reset_index()
        )

        df_daily["name"] = station_id
        df_daily = df_daily.rename(
            columns={
                "timestamp": "date",
                HS_COL: "hs",
                SWE_COL: "swe_obs"
            }
        )

        all_data.append(df_daily)

    if len(all_data) == 0:
        raise ValueError("No SMET files found.")

    out = pd.concat(all_data, ignore_index=True)
    out["date"] = pd.to_datetime(out["date"])
    out = out.sort_values(["name", "date"]).reset_index(drop=True)

    return out


##############################################################################
# PREPARE FIT DATA
##############################################################################

def prepare_fit_data(df_all, start_month=8):
    fit_parts = []

    for station, df_station in df_all.groupby("name"):
        df_station = df_station.sort_values("date").copy()
        years = np.sort(df_station["date"].dt.year.unique())

        print(f"Processing station: {station}")

        for y in years[:-1]:
            start = pd.Timestamp(year=int(y), month=8, day=1)
            end = pd.Timestamp(year=int(y) + 1, month=7, day=31)

            winter = df_station[
                (df_station["date"] >= start) &
                (df_station["date"] <= end)
            ].copy()

            if winter.empty:
                continue

            if not is_valid_winter(winter):
                print(f"  skipping winter {y}/{y+1}")
                continue

            winter["block"] = int(y)
            fit_parts.append(winter)

    if len(fit_parts) == 0:
        raise ValueError("No valid winters found for fitting.")

    fit_data = pd.concat(fit_parts, ignore_index=True)
    return fit_data


##############################################################################
# RUN DELTASNOW FOR ONE STATION/BLOCK (with warning suppression)
##############################################################################

def run_one_block(df_block, par_real):
    # Suppress FutureWarnings inside this worker process
    with warnings.catch_warnings():
        warnings.filterwarnings("ignore", category=FutureWarning)

        df_block = df_block.sort_values("date").copy()

        full_dates = pd.date_range(
            start=df_block["date"].min(),
            end=df_block["date"].max(),
            freq="D"
        )

        joined = (
            df_block[["date", "hs", "swe_obs"]]
            .set_index("date")
            .reindex(full_dates)
            .rename_axis("date")
            .reset_index()
        )

        # SnowToSwe requires no NaNs and first value == 0
        hs_series = (
            joined["hs"]
            .interpolate(method="linear", limit_direction="both")
            .clip(lower=0.0)
        )
        if hs_series.isna().any():
            return pd.DataFrame()  # skip block if interpolation still leaves NaN
        hs_arr = hs_series.to_numpy(dtype=float)
        hs_arr[0] = 0.0

        try:
            with contextlib.redirect_stdout(io.StringIO()):
                model = SnowToSwe(
                    rho_max=par_real[0],
                    rho_null=par_real[1],
                    c_ov=par_real[2],
                    k_ov=par_real[3],
                    k=par_real[4],
                    tau=par_real[5],
                    eta_null=par_real[6],
                )
            swe_mod = model.convert_list(hs_arr.tolist(), timestep=24)
            joined["swe_mod"] = np.asarray(swe_mod, dtype=float)
        except Exception:
            joined["swe_mod"] = np.nan

        joined = joined.dropna(subset=["hs", "swe_obs", "swe_mod"]).copy()
        return joined


##############################################################################
# OBJECTIVE FUNCTION
##############################################################################

def objective(par_scaled, data, par_scale, verbose=False):
    par_real = np.asarray(par_scaled) * np.asarray(par_scale)

    groups = [g for _, g in data.groupby(["name", "block"], sort=False)]

    results = Parallel(n_jobs=N_JOBS)(
        delayed(run_one_block)(g, par_real) for g in groups
    )

    results = [r for r in results if r is not None and not r.empty]

    if len(results) == 0:
        return 1e12

    dff = pd.concat(results, ignore_index=True)

    score = combined_score(dff, eps=EPS)

    if not np.isfinite(score):
        return 1e12

    if verbose:
        rmse_swe = rmse(dff["swe_obs"], dff["swe_mod"])
        rho_obs = np.where(dff["hs"] > EPS, dff["swe_obs"] / dff["hs"], np.nan)
        rho_mod = np.where(dff["hs"] > EPS, dff["swe_mod"] / dff["hs"], np.nan)
        rmse_rho = rmse(rho_obs, rho_mod)
        bias_swe = np.nanmean(dff["swe_mod"] - dff["swe_obs"])

        print(
            f"| bias_swe = {bias_swe:.4f} "
            f"| rmse_swe = {rmse_swe:.4f} "
            f"| rmse_density = {rmse_rho:.4f} "
            f"| final_score = {score:.4f}"
        )

    return float(score)


##############################################################################
# START VALUES AND BOUNDS (PHYSICAL, THEN SCALED)
##############################################################################

par_delta = np.array([
    401,
    81,
    5.1e-4,
    0.38,
    0.03,
    0.024,
    8.5e6,      
    ], dtype=float)

par_scale = np.array([
    1000,
    1000,
    0.001,
    1,
    0.1,
    0.1,
    1e7
], dtype=float)

# Physical (unscaled) lower and upper bounds
lower_unscaled = np.array([200,  20,   1e-6, 0,   0,   0,   1e5])
upper_unscaled = np.array([600, 200,   0.01, 1,   0.5, 1,   1e8])

# Scale bounds
lower_scaled = lower_unscaled / par_scale
upper_scaled = upper_unscaled / par_scale

# Scaled start values
par_start = par_delta / par_scale

print("\nScaled start parameters:")
print(par_start)

print("\nUnscaled start parameters:")
print(par_start * par_scale)

print("\nScaled bounds (lower, upper):")
for i, (l, u) in enumerate(zip(lower_scaled, upper_scaled)):
    print(f"  param_{i}: {l:.6f} -> {u:.6f}")

##############################################################################
# MAIN
##############################################################################

df_all = load_all_smet_data(INPUT_DIR)
fit_data = prepare_fit_data(df_all, start_month=SEASON_START_MONTH)

print("\nTesting objective function ...")
test_score = objective(par_start, fit_data, par_scale, verbose=True)
print("Initial score:", test_score)

print("\nStarting Differential Evolution optimization ...")

# DE control parameters
popsize = 10 * len(par_start)   # population size
maxiter = 200                   # maximum number of generations
mutation = (0.5, 1.0)           # mutation factor (dithering)
recombination = 0.7             # crossover probability

# Run differential evolution
result = differential_evolution(
    func=objective,
    bounds=list(zip(lower_scaled, upper_scaled)),
    args=(fit_data, par_scale, False),
    maxiter=maxiter,
    popsize=popsize,
    mutation=mutation,
    recombination=recombination,
    seed=123,                   # for reproducibility
    disp=True,                  # print progress every generation
    workers=1                   # DE internal parallelisation is off (we use joblib inside objective)
)

##############################################################################
# RESULTS
##############################################################################

best_par_scaled = result.x
best_par_real = best_par_scaled * par_scale

param_names = ["rho_max", "rho_null", "c_ov", "k_ov", "k", "tau", "eta_null"]
best_params = pd.Series(best_par_real, index=param_names, name="best_value")

print("\n" + "="*60)
print("OPTIMIZATION FINISHED")
print("="*60)
print("\nBest parameters (unscaled):")
print(best_params)

print("\nFinal score (objective value):")
print(result.fun)

print("\nOptimizer message:")
print(result.message)

print("\nNumber of function evaluations:")
print(result.nfev)

# Optional: print a Python call example with the optimized parameters
fmt = lambda x: f"{x:.10f}"
print("\n# Example call to SnowToSwe with optimized parameters:")
print(f"model = SnowToSwe(")
print(f"    rho_max={fmt(best_par_real[0])},")
print(f"    rho_null={fmt(best_par_real[1])},")
print(f"    c_ov={fmt(best_par_real[2])},")
print(f"    k_ov={fmt(best_par_real[3])},")
print(f"    k={fmt(best_par_real[4])},")
print(f"    tau={fmt(best_par_real[5])},")
print(f"    eta_null={fmt(best_par_real[6])}")
print(f")")